woocommerce/woocommerce-android
# [REST API] OTP login functionality for Jetpack connection

Closes: #8397


### Description

This adds the actual functionality in the OTP screen:
1. Installation process is continued if entered OTP is correct,
2. OTP via SMS request button is working,
3. Error handling and basic validation is added.


### Testing instructions
1. Login by entering a self-hosted Woo site with no Jetpack, then using site credentials,
4. On the My Store screen, tap the Jetpack banner at the bottom,
5. tap "Log In to Continue",

At this point, prepare an account that has 2FA via code, and has 2FA via SMS.

*code case*
1. Enter a wpcom account email address that has a 2FA via app enabled, tap Install Jetpack,
2. On the next screen, enter the password, then tap Install Jetpack,
3. the OTP screen will be shown. Enter OTP code from authenticator app, then tap Install Jetpack,
6. Verify that the Installation process is then continued.

*SMS case*
1. Enter a wpcom account email address that has a 2FA via SMS enabled, tap Install Jetpack,
2. On the next screen, enter the password, then tap Install Jetpack,
3. the OTP screen will be shown. Enter OTP code from SMS, then tap Install Jetpack,
4. Verify that the Installation process is then continued.

*Error case*
1. Try entering the wrong OTP code,
2. Try tapping the SMS request button for the account that has 2FA via app,
3. Turn on Airplane mode on the OTP screen, and try the buttons


### Images/gif

The screenshots are from a site that already has Jetpack installed but not connected, hence why it says "Connect Jetpack". If there's no Jetpack plugin installed, it will say "Install Jetpack".

| main | entering code | sms request dialog | login dialog |
|-|-|-|-|
| ![Screenshot_20230303_202648](https://user-images.githubusercontent.com/266376/222731911-6b8b01c6-4e2a-4c30-8033-4530005ab626.png) | ![Screenshot_20230303_202743](https://user-images.githubusercontent.com/266376/222732057-c804f743-bf25-4d32-99d8-fc408e4995ef.png) | ![Screenshot_20230303_204110](https://user-images.githubusercontent.com/266376/222734882-a0d1f597-3c46-4c3b-b517-28e5ac5c06f9.png) |  ![Screenshot_20230303_204227](https://user-images.githubusercontent.com/266376/222735258-5de37e74-10ad-42ea-8a02-d6b4aba5e7ca.png) |
